src/Illuminate/Database/Eloquent/Builder.php

```diff
@@ -428,51 +429,79 @@ public function findOrNew($id, $columns = ['*'])
+
+        if (! is_null($instance = $this->where($rawAttributes)->first())) {
             return $instance;
         }
 
-        return $this->newModelInstance(array_merge($attributes, $values));
+        return $this->newModelInstance(
+            array_merge($rawAttributes, $rawValues),
+            $this->shouldForceAttributes($attributes, $values)
+        );
     }
 
     /**
      * Get the first record matching the attributes or create it.
      *
-     * @param  array  $attributes
-     * @param  array  $values
+     * @param  array|\Illuminate\Contracts\Support\ValidatedData  $attributes
+     * @param  array|\Illuminate\Contracts\Support\ValidatedData  $values
      * @return \Illuminate\Database\Eloquent\Model|static
      */
-    public function firstOrCreate(array $attributes = [], array $values = [])
+    public function firstOrCreate($attributes = [], $values = [])
     {
-        if (! is_null($instance = $this->where($attributes)->first())) {
+        $rawAttributes = $attributes instanceof ValidatedData ? $attributes->toArray() : $attributes;
+        $rawValues = $values instanceof ValidatedData ? $values->toArray() : $values;
+
+        if (! is_null($instance = $this->where($rawAttributes)->first())) {
             return $instance;
         }
 
-        return tap($this->newModelInstance(array_merge($attributes, $values)), function ($instance) {
-            $instance->save();
-        });
+        $instance = $this->newModelInstance(
+            array_merge($rawAttributes, $rawValues),
+            $this->shouldForceAttributes($attributes, $values)
+        );
+
+        $instance->save();
+
+        return $instance;
     }
 
     /**
      * Create or update a record matching the attributes, and fill it with values.
      *
-     * @param  array  $attributes
-     * @param  array  $values
+     * @param  array|\Illuminate\Contracts\Support\ValidatedData  $attributes
+     * @param  array|\Illuminate\Contracts\Support\ValidatedData  $values
      * @return \Illuminate\Database\Eloquent\Model|static
      */
-    public function updateOrCreate(array $attributes, array $values = [])
+    public function updateOrCreate($attributes, $values = [])
     {
         return tap($this->firstOrNew($attributes), function ($instance) use ($values) {
             $instance->fill($values)->save();
         });
     }
 
+    /**
+     * Determines if the passed attributes are validated data.
+     *
+     * @param  array|\Illuminate\Contracts\Support\ValidatedData  $attributes
+     * @param  array|\Illuminate\Contracts\Support\ValidatedData  $values
+     * @return bool
+     */
+    protected function shouldForceAttributes($attributes, $values)
+    {
+        return $attributes instanceof ValidatedData && $values instanceof ValidatedData ||
+            ($attributes instanceof ValidatedData && empty($values)) ||
+            ($values instanceof ValidatedData && empty($attributes));
```

Comment:
We could address some of the concerns brought up by only forcing validated data if `fillable` and `guarded` have not been set for this model. Something like:

```suggestion
        return $this->fillable === []
            && $this->guarded === ['*']
            && ($attributes instanceof ValidatedData || empty($attributes))
            && ($values instanceof ValidatedData || empty($values);
```